docs/adrs/002-Retry_endpoint.md

```diff
@@ -0,0 +1,81 @@
+# Retry Endpoint
+
+Date: 2021-04-27
+
+The current requirement for the Batch/Bulk UI is to be able to retry jobs. Using a job id, it should be possible to get information from the database.
+
+The current implementation of retry is to run jobs using the `run_job` or `run_job_batch` endpoint. This is not adequate due to the following deficiencies:
+
+
+* Lack of book-keeping for the parent job and the child job relationship: Launching a new job will not update the relationship between those jobs
+* Lack of book-keeping for number of retries for a particular job / set of job inputs
+* Lack of ability to launch multiple jobs using the `run_job_batch` endpoint without creating a new parent job
+* Lack of ability to ensure that the proper catalog version /git commit of an app is used from the front end based on a tag, such as "beta/dev/release"
+* Lack of ability to specify which retries succeeded and which ones failed during submit time. 
+
+* Code is split more than is necessary
+
+## Author(s)
+
+@bio-boris
+
+## Status
+
+Pending
+
+## Alternatives Considered
+
+* Not book-keeping, or doing minimal book-keeping and calling run_job multiple times
+* Re-writing run_job/run_job_batch to address the aforementioned deficiencies
+* Creating a retry endpoint dedicated to addressing book-keeping and job launching features
+
+## Decision Outcome
+
+Pending
+
+* Link to spec file with inputs and outputs for the retry endpoint goes here
+* Link to Jira Ticket with business logic documentation for success and cancel cases
+
+### Possible additional things to think about
+* Creating sets of objects and what to do at the end of a batch run
+* What to do about a set if a child task fails during processing 
+* Convenience endpoints that operate on the parent_job_id or list of child job ids may be out of scope (e.g. cancel all jobs with a certain status)
+
+## Consequences
+
+Pending decision
+
+## Pros and Cons of the Alternatives
+
+### Not book-keeping, or doing minimal book-keeping and calling `run_job` multiple times
+* `+` Can re-use existing endpoints without any additional work
+* `+` Less api endpoints to manage
+* `-` Issues on re-rendering/regenerating a cell based on just the job record
```

Comment:
can you explain this a bit more?

Reply:
We can in theory attempt to re-create a cell solely by the job record.

Reply:
if all the child job IDs have been saved, this would not be a problem.

Reply:
RIght now run_job as it currently stands, doesn't save number of retries for a job, update the parent job,  and the narrative wouldn't know which jobs have been retried, and thus to show a row for, and which ones not to. Thus it would either not render the retried jobs because it doesn't know about them, or if all of them are append to the list of child jobs, it would render all of the retried jobs as well.   The purpose of this section is to explain some deficiencies  with using run_job as it is today.

Reply:
This is really just the same point as is mentioned below, that whatever the client is (the narrative, some third-party app, whatever) could do a lot of record keeping about the relationships between the different jobs; if it didn't, there wouldn't be a way to reconstruct the job relationships just using ee2 data. I think I would say:

* `-` Narrative would be forced to maintain detailed job records to be able to re-render or regenerate a batch job cell where jobs within the batch had been retried.